### PR TITLE
Fix linear memory page count issues

### PR DESCRIPTION
**Pull request by wenyongh**
Fix issue reported in #1289 and #1371.
Enable to set the max page count to 65536.

### CODE REVIEW

File: core/iwasm/aot/aot_runtime.c [hunk 1]
```diff
@@ -2062,94 +2070,103 @@ aot_enlarge_memory(AOTModuleInstance *module_inst, uint32 inc_page_count)
...
+        total_size_new = UINT32_MAX;
     }
 
 #if WASM_ENABLE_SHARED_MEMORY != 0
-    if (memory_inst->is_shared) {
-        /* For shared memory, we have reserved the maximum spaces during
-            instantiate, only change the cur_page_count here */
-        memory_inst->cur_page_count = total_page_count;
+    if (memory->is_shared) {
+        memory->num_bytes_per_page = UINT32_MAX;
```
Comment (lum1n0us) [inline core/iwasm/aot/aot_runtime.c:2083]: If the code didn't go in the block between L2075 and L2078, do we still have to change `num_bytes_per_page`?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 2]
```diff
@@ -208,7 +208,7 @@ memory_instantiate(WASMModuleInstance *module_inst, uint32 num_bytes_per_page,
         num_bytes_per_page += heap_size;
         if (num_bytes_per_page < heap_size) {
             set_error_buf(error_buf, error_buf_size,
-                          "memory size must be at most 65536 pages (4GiB)");
+                          "failed to insert app heap into linear memory");
```
Comment (lum1n0us) [inline core/iwasm/interpreter/wasm_runtime.c:211]: shall we append something like "maybe use --heap_size=0 and try again"?

**Review by lum1n0us — COMMENTED** [review]
a little comments

File: core/iwasm/aot/aot_runtime.c (hunk 1 above)
Comment (wenyongh) [reply core/iwasm/aot/aot_runtime.c:2092]: No, only when the memory data size is out of range, or, it reaches 4GB (65536 pages * 65536 bytes_each_page), we need to resize it. As the type of aot_inst->memory_data_size is uint32, and its range is 0 to 4GB-1, and we had better keep the AOT ABI unchanged.

File: core/iwasm/interpreter/wasm_runtime.c (hunk 2 above)
Comment (wenyongh) [reply core/iwasm/interpreter/wasm_runtime.c:211]: OK, let's add more prompt.

**Comment by lum1n0us** [comment]
LGTM

**Pull request merged**